### PR TITLE
DKG Storage Pt. 1: Update storage layer

### PR DESCRIPTION
**Pull request by jordanschalm**
This PR is Part I of an update to DKG State storage.
* refactors DKG key storage to store start and end states
* adds the `SafeBeaconPrivateKeys` interface which provides a safe way for signing components to access beacon keys
* updates storage-layer tests

Pt II (https://github.com/onflow/flow-go/pull/1721) updates all components depending on the modified storage components.
Pt III (https://github.com/onflow/flow-go/pull/1732) updates the beacon key checking to occur atomically with the transition to `EpochCommitted` phase.

### CODE REVIEW

File: storage/badger/dkg_state.go [hunk 1]
```diff
@@ -0,0 +1,144 @@
+package badger
```
Comment (kc1116) [inline storage/badger/dkg_state.go:1]: Can you add godoc comments for the exported funcs

File: storage/badger/dkg_state.go [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+
+func (ds *DKGState) SetDKGEndState(epochCounter uint64, endState flow.DKGEndState) error {
+	return ds.db.Update(operation.InsertDKGEndStateForEpoch(epochCounter, endState))
+}
+
+type SafeBeaconPrivateKeys struct {
+	state *DKGState
+}
+
+func NewSafeBeaconPrivateKeys(state *DKGState) (*SafeBeaconPrivateKeys, error) {
```
Comment (kc1116) [inline storage/badger/dkg_state.go:106]: remove the error returned from this func

File: storage/badger/operation/dkg_test.go [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/storage"
+	"github.com/onflow/flow-go/utils/unittest"
+)
+
+// TestInsertMyDKGPrivateInfo_StoreRetrieve tests writing and reading private DKG info.
+func TestMyBeaconPrivateKey_StoreRetrieve(t *testing.T) {
+	unittest.RunWithBadgerDB(t, func(db *badger.DB) {
+
+		t.Run("when not stored", func(t *testing.T) {
```
Comment (kc1116) [inline storage/badger/operation/dkg_test.go:20]:
```suggestion
		t.Run("should return error not found when not stored", func(t *testing.T) {
```

**Review by kc1116 — APPROVED** [review]
LGTM - minor comments

File: storage/badger/dkg_state_test.go [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+
+			// store a key
+			expected := unittest.RandomBeaconPriv().PrivateKey
+			err := dkgState.InsertMyBeaconPrivateKey(epochCounter, expected)
+			assert.NoError(t, err)
+			// mark dkg unsuccessful
+			err = dkgState.SetDKGEndState(epochCounter, flow.DKGEndStateInconsistentKey)
+			assert.NoError(t, err)
+
+			key, safe, err := safeKeys.RetrieveMyBeaconPrivateKey(epochCounter)
```
Comment (zhangchiqing) [inline storage/badger/dkg_state_test.go:128]: 👍

**Review by zhangchiqing — APPROVED** [review]
Looks good

File: storage/dkg.go [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+// SafeBeaconKeys is a safe way to access beacon keys.
+type SafeBeaconKeys interface {
+
+	// RetrieveMyBeaconPrivateKey retrieves my beacon private key for the given
+	// epoch, only if my key has been confirmed valid and safe for use.
+	//
+	// Returns:
+	// * (key, true, nil) if the key is present and confirmed valid
+	// * (nil, false, nil) if the key has been marked invalid (by SetDKGEnded)
```
Comment (zhangchiqing) [inline storage/dkg.go:51]:
```suggestion
	// * (nil, false, nil) if no key was generated or the key has been marked invalid (by SetDKGEnded)
```

File: storage/badger/dkg_state.go [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+		val, err := ds.keyCache.Get(epochCounter)(tx)
+		if err != nil {
+			return nil, err
+		}
+		return val.(*encodable.RandomBeaconPrivKey), nil
+	}
+}
+
+func (ds *DKGState) InsertMyBeaconPrivateKey(epochCounter uint64, key crypto.PrivateKey) error {
+	encodableKey := &encodable.RandomBeaconPrivKey{PrivateKey: key}
```
Comment (tarakby) [inline storage/badger/dkg_state.go:74]: Nice conversion to the encodable type inside the function 👌🏼

File: storage/badger/dkg_state.go [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+}
+
+func (ds *DKGState) RetrieveMyBeaconPrivateKey(epochCounter uint64) (crypto.PrivateKey, error) {
+	tx := ds.db.NewTransaction(false)
+	defer tx.Discard()
+	encodableKey, err := ds.retrieveKeyTx(epochCounter)(tx)
+	if err != nil {
+		return nil, err
+	}
+	return encodableKey.PrivateKey, nil
```
Comment (tarakby) [inline storage/badger/dkg_state.go:85]: 👌🏼

File: storage/badger/dkg_state.go [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+
+type SafeBeaconPrivateKeys struct {
+	state *DKGState
+}
+
+func NewSafeBeaconPrivateKeys(state *DKGState) (*SafeBeaconPrivateKeys, error) {
+	return &SafeBeaconPrivateKeys{state: state}, nil
+}
+
+func (sk *SafeBeaconPrivateKeys) RetrieveMyBeaconPrivateKey(epochCounter uint64) (key crypto.PrivateKey, safe bool, err error) {
```
Comment (tarakby) [inline storage/badger/dkg_state.go:110]: maybe we can avoid `sk` as we usually reserve it for private keys

**Review by tarakby — APPROVED** [review]
🏄🏼

File: storage/badger/dkg_state.go (hunk 1 above)
Comment (jordanschalm) [reply storage/badger/dkg_state.go:1]: https://github.com/onflow/flow-go/pull/1720/commits/23626c59c4dc4a91535a3b3a6c81476b442c6c4d

File: storage/badger/dkg_state.go (hunk 2 above)
Comment (jordanschalm) [reply storage/badger/dkg_state.go]: https://github.com/onflow/flow-go/pull/1720/commits/d2260476fa1f04ae0cfbaf4deb7ba1875d60abcf

File: storage/badger/dkg_state.go (hunk 8 above)
Comment (jordanschalm) [reply storage/badger/dkg_state.go]: https://github.com/onflow/flow-go/pull/1720/commits/e82aa997e797245961a3e904f4368537239c6ddc

**Pull request merged**